### PR TITLE
E2E tests

### PR DESCRIPTION
**Pull request by TheUnderScorer**
- Bootstrap `e2e-app` using expo for running the E2E tests via Detox.
- Introduced new `e2e-tests.yml` workflow for running e2e tests. It runs on on pull requests and after release. When triggered on release, it will install released version and use it for testing.
- Set `minSdkVersion` to `24` on android to fix failing E2E app build. Reflects the same change made in [react-native 0.76](https://github.com/facebook/react-native/blob/main/CHANGELOG.md#v0760)

### CODE REVIEW

File: e2e-app/README.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+npx detox build --configuration ios.sim.release
+```
+When debugging tests locally, it is recommended to use ios.sim.debug for faster build and easier logs access.
+```bash
+npx detox build --configuration ios.sim.debug
+```
+
+### 2. Running tests
+```bash
+npx detox test --configuration ios.sim.release
```
Comment (necipallef) [inline e2e-app/README.md:42]: This command requires `.env` file with 6 variables (`MINIMUM_US_DEFAULT_PUBLIC_KEY` and others) in order to operate. Maybe we should add an `.env.example`?

**Review by necipallef — CHANGES_REQUESTED** [review]

**Comment by necipallef** [comment]
Tested with Android, all is good, a few feedback:
- I tested it with a real device instead of emulator. It was easy to find what configuration to change, I just looked it up in detox docs. No need to mention how to do this in this repo because it is a good idea to direct users to read detox docs if they need to do something different, so all is good
- I faced an error saying "Could not get unknown property 'release' for SoftwareComponent container of type org.gradle.api.internal.component.DefaultSoftwareComponentContainer.". I expect that you've faced the same, probably? The fix is to make modifications in `e2e-app/node_modules/expo-modules-core/android/ExpoModulesCorePlugin.gradle` so that it doesn't throw an error. We can make this change permanent by applying the change on each `pnpm install` by using a library like [patch-package](https://www.npmjs.com/package/patch-package)
- I had to create `e2e-app/android/local.properties` file with the content of `sdk.dir=<REDACTED>` to indicate Android SDK path. I think we should document this

File: e2e-app/app.config.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    package: 'com.fingerprint.e2eapp',
+  },
+  plugins: [
+    [
+      'expo-build-properties',
+      {
+        ios: {
+          extraPods,
+        },
+        android: androidBuildProperties,
```
[the remark this answers is not in the document]
Comment (TheUnderScorer) [reply e2e-app/app.config.ts:103]: Actually, in most cases we pass `androidBuildProperties` as-is. In case of using a different react-native versions, it might be modified by `stripExtraBuildProperties` if it is defined in `reactNativeMetadata` for given RN version. 
That makes it tricky to extract as separate function, and I think it's better to leave it as it is right now.

File: e2e-app/app.config.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  },
+  android: {
+    package: 'com.fingerprint.e2eapp',
+  },
+  plugins: [
+    [
+      'expo-build-properties',
+      {
+        ios: {
+          extraPods,
```
[the remark this answers is not in the document]
Comment (TheUnderScorer) [reply e2e-app/app.config.ts]: Good point, done!

File: e2e-app/README.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+- Device running the latest MacOS version
+- Running `iPhone 16` simulator
+
+> To use different simulator, update the `devices.simulator.device.type` value in `.detoxrc.js`
+
+### 1. Building the app
+```bash
+npx detox build --configuration ios.sim.release
+```
+When debugging tests locally, it is recommended to use ios.sim.debug for faster build and easier logs access.
```
[the remark this answers is not in the document]
Comment (TheUnderScorer) [reply e2e-app/README.md]: Fixed, thanks!

File: .github/workflows/e2e-tests.yml [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+        uses: actions/upload-artifact@v4
+        with:
+          name: android-detox-artifacts-${{matrix.react-native}}
+          path: '**/detox-artifacts'
+
+      - uses: geekyeggo/delete-artifact@f275313e70c08f6120db482d7a6b98377786765b
+        name: Delete build artifact
+        if: success()
+        with:
+          name: ios-app-release-${{ matrix.react-native }}
```
[the remark this answers is not in the document]
Comment (TheUnderScorer) [reply .github/workflows/e2e-tests.yml]: Good catch, fixed!

File: e2e-app/e2e/identification.test.ts [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+const VISITOR_ID_REGEX = /^[a-zA-Z\d]{20}$/
+
+async function launchApp(params?: DeviceLaunchAppConfig) {
+  await device.launchApp(params)
+
+  await wait(4000)
+}
+
+describe.each([
+  ['us', process.env.MINIMUM_US_DEFAULT_PUBLIC_KEY, process.env.MINIMUM_US_DEFAULT_PRIVATE_KEY],
```
[the remark this answers is not in the document]
Comment (TheUnderScorer) [reply e2e-app/e2e/identification.test.ts:24]: It's naming that we use in our [dx-team-orchestra](https://github.com/fingerprintjs/dx-team-orchestra/blob/main/projects/conductor/utils/testData.ts) project. `MINIMUM` indicates sub with the least amount of features available.

File: e2e-app/README.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+The app is intended to run via Detox. To learn more, head to the sections below.
+
+## Running iOS tests
+
+### Prerequisites
+- [React-Native Prerequisites](https://reactnative.dev/docs/next/set-up-your-environment?platform=ios)
+- [Detox Prerequisites](https://wix.github.io/Detox/docs/introduction/environment-setup)
+- Device running the latest MacOS version
+- Running `iPhone 16` simulator
```
[the remark this answers is not in the document]
Comment (TheUnderScorer) [reply e2e-app/README.md:27]: Thanks for mentioning this. I couldn't find any mentions about it in react-native, expo or detox docs, so I think it's better to not add it here, in order to keep the steps simple, and since this seems more like an edge-case.

File: e2e-app/README.md (hunk 1 above)
Comment (TheUnderScorer) [reply e2e-app/README.md:42]: Done!

File: e2e-app/README.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+### 2. Running tests
+```bash
+npx detox test --configuration ios.sim.release
+```
+
+The configuration passed via `--configuration` must match the one used in `detox build`.
+
+> If you're using `ios.sim.debug` configuration, don't forget to run `yarn start` to start the bundler before running tests.
+
+## Running android tests
```
[the remark this answers is not in the document]
Comment (TheUnderScorer) [reply e2e-app/README.md]: Fixed

File: e2e-app/app.config.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (necipallef) [reply e2e-app/app.config.ts:103]: For instance, this is the equivalent code:
```
import { ExpoConfig } from '@expo/config'
import { dependencies } from './package.json'
import { ExtraIosPodDependency } from 'expo-build-properties/build/pluginConfig'
import * as semver from 'semver'

let newArch = false
const disableNewArch = () => {
  newArch = false
}

function createAndroidBuildProperties() {
  let rawReactNativeVersion = dependencies['react-native']

  // Ensure that react-native version contains patch range so that semver can parse it
  if (rawReactNativeVersion.split('.').length === 2) {
    rawReactNativeVersion = rawReactNativeVersion + '.0'
  }

  const rnVersion = semver.parse(rawReactNativeVersion)
  if (!rnVersion) {
    throw new Error(`Could not parse react-native version: ${dependencies['react-native']}`)
  }

  const androidBuildProperties = {
    minSdkVersion: 24,
    compileSdkVersion: 35,
    targetSdkVersion: 35,
    buildToolsVersion: '34.0.0',
    extraMavenRepos: ['https://maven.fpregistry.io/releases'],
  } as Record<string, any>

  const stripExtraBuildProperties = () => {
    delete androidBuildProperties.compileSdkVersion
    delete androidBuildProperties.targetSdkVersion
    delete androidBuildProperties.buildToolsVersion
  }

  const metadataKey = rnVersion.major + '.' + rnVersion.minor

  /**
   * Holds metadata for specific versions of React Native, including callbacks
   * for version-specific adjustments or operations.
   *
   * @type {Record<string, { callbacks: (() => void)[] }>}
   *
   * - The key represents the version of React Native.
   * - The value is an object containing:
   *   - `callbacks`: An array of functions to execute in the context of the respective React Native version.
   */
  const reactNativeMetadata: Record<string, { callbacks: (() => void)[] }> = {
    0.73: {
      callbacks: [disableNewArch, stripExtraBuildProperties],
    },
    0.74: {
      callbacks: [disableNewArch, stripExtraBuildProperties],
    },
    0.75: {
      callbacks: [disableNewArch, stripExtraBuildProperties],
    },
    0.76: {
      callbacks: [stripExtraBuildProperties],
    },
  }

  if (reactNativeMetadata[metadataKey]?.callbacks?.length) {
    reactNativeMetadata[metadataKey].callbacks.forEach((callback) => callback())
  }

  return androidBuildProperties
}

function createExtraPods() {
  const extraPods = [] as ExtraIosPodDependency[]
  const sdkVersion = dependencies['@fingerprintjs/fingerprintjs-pro-react-native']
  const usesLocalPackage = sdkVersion === '../'
  if (usesLocalPackage) {
    extraPods.push({
      name: 'RNFingerprintjsPro',
      path: '../node_modules/@fingerprintjs/fingerprintjs-pro-react-native',
    })
  }

  return extraPods
}

const config: ExpoConfig = {
  name: 'E2EApp',
  slug: 'e2eapp',
  version: '1.0.0',
  orientation: 'portrait',
  icon: './assets/images/icon.png',
  scheme: 'e2eapp',
  newArchEnabled: newArch,
  splash: {
    backgroundColor: '#ffffff',
  },
  ios: {
    bundleIdentifier: 'com.fingerprint.e2eapp',
  },
  android: {
    package: 'com.fingerprint.e2eapp',
  },
  plugins: [
    [
      'expo-build-properties',
      {
        ios: {
          extraPods: createExtraPods(),
        },
        android: createAndroidBuildProperties(),
      },
    ],
    '@config-plugins/detox',
    './plugins/withGradleProperties.js',
  ],
}

export default config

```

of course it needs some polishing, but as-is, would this not work? I'd prefer this over current because it is easier to read, reader reads the `createAndroidBuildProperties` function name first, then reads the implementation as opposed to reading implementation right away without knowing what the code is trying to do.

File: e2e-app/README.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
+# E2E-App
+
+This application is used for running E2E tests via [Detox](https://github.com/wix/Detox)
+and was bootstraped using [Expo](https://expo.dev).
```
Comment (necipallef) [inline e2e-app/README.md:4]: typo `bootstraped` -> `bootstrapped`

File: e2e-app/README.md [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+## Get started
+
```
Comment (necipallef) [inline e2e-app/README.md:7]: let's mention here that we need to build the SDK before doing any of the steps below because it may not be obvious

**Review by necipallef — APPROVED** [review]
